go/controller/translator/adk_api_translator.go

```diff
@@ -0,0 +1,730 @@
+	objMeta := metav1.ObjectMeta{
+		Name:        agent.Name,
+		Namespace:   agent.Namespace,
+		Annotations: agent.Annotations,
+		Labels:      newLabels,
+	}
+	if agent.Spec.Deployment == nil {
+		spec := defaultDeploymentSpec(objMeta.Name)
+		outputs.Deployment = &appsv1.Deployment{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "apps/v1",
+				Kind:       "Deployment",
+			},
+			ObjectMeta: objMeta,
+			Spec: appsv1.DeploymentSpec{
+				Replicas: spec.Replicas,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"app":     "kagent",
+						"kagent":  agent.Name,
+						"version": "v1alpha1",
+					},
+				},
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: objMeta,
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{spec.Container},
+						Volumes:    spec.Volumes,
+					},
+				},
+			},
+		}
+	} else {
+		return nil, fmt.Errorf("We need to figure out merging the deployment spec")
+	}
+
+	outputs.Service = &corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Service",
+		},
+		ObjectMeta: objMeta,
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				"app":     "kagent",
+				"kagent":  agent.Name,
+				"version": "v1alpha1",
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "http",
+					Port:       8080,
+					TargetPort: intstr.FromInt(8080),
+				},
+			},
+			Type: corev1.ServiceTypeClusterIP,
+		},
+	}
+
+	outputs.ConfigMap = &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: objMeta,
+		Data:       map[string]string{},
+	}
+
+	if err := controllerutil.SetControllerReference(agent, outputs.Deployment, a.kube.Scheme()); err != nil {
+		return nil, err
+	}
+
+	if err := controllerutil.SetControllerReference(agent, outputs.Service, a.kube.Scheme()); err != nil {
+		return nil, err
+	}
+
+	if err := controllerutil.SetControllerReference(agent, outputs.ConfigMap, a.kube.Scheme()); err != nil {
+		return nil, err
+	}
+
+	return outputs, nil
+}
+
+func defaultDeploymentSpec(name string) *v1alpha1.DeploymentSpec {
+	// TODO: Come up with a better way to do tracing config for the agents
+	envVars := slices.Collect(utils.Map(
+		utils.Filter(
+			slices.Values(os.Environ()),
+			func(envVar string) bool {
+				return strings.HasPrefix(envVar, "OTEL_")
+			},
+		),
+		func(envVar string) corev1.EnvVar {
+			parts := strings.SplitN(envVar, "=", 2)
+			return corev1.EnvVar{
+				Name:  parts[0],
+				Value: parts[1],
+			}
+		},
+	))
+
+	envVars = append(envVars, corev1.EnvVar{
+		Name: "KAGENT_NAMESPACE",
+		ValueFrom: &corev1.EnvVarSource{
+			FieldRef: &corev1.ObjectFieldSelector{
+				FieldPath: "metadata.namespace",
+			},
+		},
+	})
+
+	return &v1alpha1.DeploymentSpec{
+		Replicas: ptr.To(int32(1)),
+		Container: corev1.Container{
+			Name:            "kagent",
+			Image:           fmt.Sprintf("cr.kagent.dev/kagent-dev/kagent/app:%s", version.Get().Version),
```

Comment:
can the prefix be configurable, for people mirroring images?

Reply:
Definitely, mind if I add in a follow-up?